### DISCUSSION TITLE
Cannot save pdf file

### DISCUSSION THREAD

**Discussion by pmi123** [pdfarranger/pdfarranger#1017] [Q&A] [answered]
Program will not save pages
----------------

Problem
------------------------
I imported 2 pdf files. Arranged the pages as I needed them. Tried to save the combined pdf document and got this error
Use pike.PDF.pages interface to copy pages from one PDf to another.

Expected behavior
-----------------------------
To be able to save the combined pdf file.

Screenshots
-----------

![Screenshot from 2023-12-21 13-00-44](https://github.com/pdfarranger/pdfarranger/assets/3002127/60a8a819-4399-4dca-876e-bc7641c5fdb5)

System and Versions
-------------------

-   PDF Arranger version 1.8.2 (distro package
-   pikepdf version 8.10.1 (distro package)
-   OS name and version Ubuntu 22.04

**Comment by jeromerobert** [comment]
*pikepdf version 8.10.1 (distro package)* <-- No. pikepdf 8.10.1 is not yet packaged for any Debian / Ubuntu distro.

Anyway PDF Arranger 1.8.2 will not work with 8.10 as its by far too old. You need PDF Arranger 1.10.1 if you want to use PikePDF 8.10. If you want to keep usig PDF Arranger 1.8.2, use the pikepdf version which packaged in Ubuntu 22 (i.e. pikepdf 5)

For the record the full stack of this bug is:

```
Traceback (most recent call last):
  File "/home/robert/devel/pdfarranger/pdfarranger/pdfarranger.py", line 976, in choose_export_pdf_name
    self.save(mode, file_out)
  File "/home/robert/devel/pdfarranger/pdfarranger/pdfarranger.py", line 160, in wrapper
    func(*args, **kwargs)
  File "/home/robert/devel/pdfarranger/pdfarranger/pdfarranger.py", line 1050, in save
    exporter.export(self.pdfqueue, to_export, file_out, mode, m)
  File "/home/robert/devel/pdfarranger/pdfarranger/exporter.py", line 177, in export
    new_page = pdf_output.copy_foreign(current_page)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
NotImplementedError: Use pikepdf.Pdf.pages interface to copy pages from one PDF to another.
```

**Comment by pmi123** [comment]
I did a sudo apt update and sudo apt upgrade and pdf arranger does not change version. How do I update it?

**Comment by jeromerobert** [comment]
You cannot upgrade it using apt because only 1.8.2 is package in ubuntu 22. You can either install it with snap or flatpak or from source (see the README).

But IMHO you should downgrade pikepdf instead. That is you should find why and how you got a so recent version of pikepdf.

**Comment by pmi123** [comment]
I am confused. 

When I open PDF Arranger and check the about box this is what I see
![Screenshot from 2023-12-22 18-05-01](https://github.com/pdfarranger/pdfarranger/assets/3002127/4c873bb6-ecc9-4228-b5da-e57bf6642989)
which implies my version of pdf arranger is using pikepdf 8.10.1.

However, when I check my Ubuntu 22.04 systems for the version of pikepdf, I see this
ii  python3-pikepdf 5.0.1+dfsg-1 amd64        Python library to read and write PDFs with QPDF

which shows version 5.0.1. 

What do I do now? I want pdf arranger to work as it should.

**Comment by dreua** [comment]
You probably have multiple versions of pikepdf installed, did you install anything with pip which could have brought it as a dependency?
Using the flatpak is the fast workaround, almost guaranteed to work ;)

**Comment by dreua** [comment, marked as answer]
Try `pip uninstall pikepdf` and see what that would do. Note that in most cases pip shouldn't be used as root, so only do `sudo pip anything` when you know what you're doing.

You might need to check the "site-packages" directories manually for pikepdf packages, you can list those locations via `python -m site`.

**Comment by pmi123** [comment]
Thanks! Somehow pikepdf was installed using sudo pip. Probably because I am
so used to sudo apt that I unconsciously typed sudo pip.

I removed pikepdf, uninstalled pdfarranger using sudo apt, and then
reinstalled pdfarranger using sudo apt, and it works as I expect it to.

Cheers!

**Comment by fmpanico** [comment]
I'm on Ubuntu 22.04.04 and I was facing the same issue, having installed pdfarranger 1.8.2 via command line apt install.

Removing it (with sudo apt remove --purge pdfarranger) and reinstalling via snap (sudo snap install pdfarranger) I got version 1.10.1 and fixed the issue.

**Comment by gregreen** [comment]
So as things stand, the pdfarranger apt package is broken in Ubuntu? Is there any plan to fix this?

**Reply by kbengs** [reply]
You too have Ubuntu 22.04 and pdfarranger 1.8.2 and pikepdf 8.10.1 ?
The apt package is by the way not built by anyone here.

**Reply by dreua** [reply]
Btw, nothing here suggests that the apt package is broken, it is just users environments or installations that are broken for a different reason. If you are unwilling to learn about your system or ask good questions please just use the flatpaks (or snaps if you prefer that). They are designed to work most of the time, even in otherwise broken or polluted environments.

**Reply by gregreen** [reply]
`apt install pdfarranger` yields a broken install on Ubuntu 22.04 LTS. I'm sorry, but that's the definition of "broken."

Snaps and flatpaks are indeed an alternative (though a very bloated one), but the apt package is broken.

**Reply by dreua** [reply]
*sigh* I just wasted some time spinning up a fresh Ubuntu 22.04 lts machine, activating "uinverse" and testing the PDF Arranger package there. It worked just fine, therefore it is not "broken", at least in my definition of that word.
That is why I assume that something in your environment is broken or at least responsible for PDF Arranger being not functional. Anyway, you are still complaining to the wrong audience here, as kbengs already pointed out. Please find a way to contact the maintainers of the package on Ubuntu and discuss it with them.

**Reply by dreua** [reply]
If you do reach the Ubuntu maintainers please tell them to add a `-s` flag to the interpreter call in `/usr/bin/pdfarranger` (shebang in the first line), this should fix this, I just figured out that this happens automatically on Fedora. `-sP` would be even better as soon as it is supported by the system's Python interpreter.

**Reply by gregreen** [reply]
Well, on my fairly typical Ubuntu system, apt yields a broken install. If the apt install route fails for me, it's going to fail for a lot of people. Anyways, I've already solved my use case with ghostscript, so I'll let someone else track down the Ubuntu maintainers to let them know about the issues with pdfarranger in Ubuntu 22.04.

**Reply by dreua** [reply]
https://bugs.launchpad.net/ubuntu/+source/pdfarranger/+bug/2077836
